packages/react-server-cli/src/startServer.js

```diff
@@ -31,69 +35,113 @@ export default (routesRelativePath, {
-// http://localhost:port/. returns a promise that resolves when the server has
-// started.
+// http://localhost:port/. returns an object with two properties, started and stop;
+// see the default function doc for explanation.
 const startHtmlServer = (serverRoutes, port) => {
-	return new Promise((resolve) => {
-		logger.info("Starting HTML server...");
+	const server = express();
+	const httpServer = http.createServer(server);
+	return {
+		stop: serverToStopPromise(httpServer),
+		started: new Promise((resolve, reject) => {
+			logger.info("Starting HTML server...");
 
-		const server = express();
-		server.use(compression());
-		reactServer.middleware(server, require(serverRoutes));
+			server.use(compression());
+			reactServer.middleware(server, require(serverRoutes));
 
-		http.createServer(server).listen(port, () => {
-			logger.info(`Started HTML server on port ${port}`);
-			resolve();
-		});
-	});
+			httpServer.on('error', (e) => {
+				console.error("Error starting up HTML server");
+				console.error(e);
+				reject(e)
+			});
+			httpServer.listen(port, (e) => {
+				if (e) {
+					reject(e);
+					return;
+				}
+				logger.info(`Started HTML server on port ${port}`);
+				resolve();
+			});
+		}),
+	};
 };
 
 // given a webpack compiler and a port, compile the JavaScript code to static
 // files and start up a web server at http://localhost:port/ that serves the
-// static compiled JavaScript. returns a promise that resolves when the server
-// has started.
+// static compiled JavaScript. returns an object with two properties, started and stop;
+// see the default function doc for explanation.
 const startStaticJsServer = (compiler, port) => {
-	return new Promise((resolve) => {
-		compiler.run((err, stats) => {
-			handleCompilationErrors(err, stats);
+	const server = express();
+	const httpServer = http.createServer(server);
+	return {
+		stop: serverToStopPromise(httpServer),
+		started: new Promise((resolve, reject) => {
+			compiler.run((err, stats) => {
+				const error = handleCompilationErrors(err, stats);
+				if (error) {
+					reject(error);
+				}
 
-			logger.debug("Successfully compiled static JavaScript.");
-			// TODO: make this parameterized based on what is returned from compileClient
-			let server = express();
-			server.use('/', compression(), express.static('__clientTemp/build'));
-			logger.info("Starting static JavaScript server...");
+				logger.debug("Successfully compiled static JavaScript.");
+				// TODO: make this parameterized based on what is returned from compileClient
+				server.use('/', compression(), express.static(`__clientTemp/build`));
+				logger.info("Starting static JavaScript server...");
 
-			http.createServer(server).listen(port, () => {
-				logger.info(`Started static JavaScript server on port ${port}`);
-				resolve();
+				httpServer.on('error', (e) => {
+					console.error("Error starting up JS server");
+					console.error(e);
+					reject(e)
```

Comment:
This looks like more robust handling of server start failure than https://github.com/redfin/react-server/pull/80.